[Fuzzilli] Ignore static initializer linker errors when building JavaScriptCore.framework

ddkilzer opened the pull request:
#### bd4957430c87e2746ba672c7fcb3f30af48e817d
<pre>
[Fuzzilli] Ignore static initializer linker errors when building JavaScriptCore.framework
&lt;<a href="https://bugs.webkit.org/show_bug.cgi?id=296950">https://bugs.webkit.org/show_bug.cgi?id=296950</a>&gt;
&lt;<a href="https://rdar.apple.com/157593751">rdar://157593751</a>&gt;

Reviewed by Elliott Williams and Yusuke Suzuki.

Internal SDKs for iOS 26 and macOS 26 enabled `-no_inits` by default for
targets that are included in the dyld shared cache, which broke the
Fuzzilli build.  This is fixed by adding linker flags to ignore static
initializers and to reassure the linker that the built product won&apos;t be
included in a dyld shared cache.

* Configurations/Sanitizers.xcconfig:
(WK_FUZZILLI_OTHER_LDFLAGS_YES):
- Add `-no_warn_inits` and `-not_for_dyld_shared_cache` to fix linking
  of JavaScriptCore.framework with Fuzzilli enabled.

Canonical link: <a href="https://commits.webkit.org/298304@main">https://commits.webkit.org/298304@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ccaa7716cdb8b210cfde4a8ae53125975c783034

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115086 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34799 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25283 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/121201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/65726 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/83ae476d-6501-4372-a1da-206ef364fc5c) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35446 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43361 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/121201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/65726 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a332d272-96f3-4bb7-bd90-86befb6bbfb4) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118034 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/28257 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/103323 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/121201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/27428 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/64854 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/97641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/21560 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/124389 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42055 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/31452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/124389 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42425 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/99513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/124389 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/41241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/19081 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/38052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18410 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/41930 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/41471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/44790 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/43206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Configurations/Sanitizers.xcconfig:
```diff
@@ -116,7 +116,7 @@ WK_LIBFUZZER_OTHER_LDFLAGS_YES = $(WK_SANITIZER_COVERAGE_OTHER_FLAGS);
 
 WK_FUZZILLI_OTHER_CFLAGS_YES = $(WK_SANITIZER_COVERAGE_OTHER_FLAGS) -DENABLE_FUZZILLI=1;
-WK_FUZZILLI_OTHER_LDFLAGS_YES = $(WK_SANITIZER_COVERAGE_OTHER_FLAGS);
+WK_FUZZILLI_OTHER_LDFLAGS_YES = $(WK_SANITIZER_COVERAGE_OTHER_FLAGS) -Xlinker -no_warn_inits -Xlinker -not_for_dyld_shared_cache;
 
```
What is the reason of this error? Can you add comment here about that?

ddkilzer replied:
I believe the linker defaults to linking with `-no_inits` in more recent (internal) SDKs for projects (or targets) that are part of the dyld shared cache, since `-no_inits` is not see anywhere on the command-line (which also means that `JavaScriptCore.xcconfig` was not setting the command-line switch).

The Fuzzilli configuration hadn't been built with an SDK since that change, so we didn't know about it until recently.

The fix limits setting these linker switches to when `ENABLE_FUZZILLI=YES`.

Constellation replied:
OK, r=me with adding these descriptions here as a comment.

ddkilzer replied:
> OK, r=me with adding these descriptions here as a comment.

I'm a little unclear on the "ask" here, but it sounds like you want me to put the above text in the commit message, so I'll do that before landing.

emw-apple reviewed (approved):
rs=me

Commit:
[Fuzzilli] Ignore static initializer linker errors when building JavaScriptCore.framework
<https://bugs.webkit.org/show_bug.cgi?id=296950>
<rdar://157593751>

Reviewed by Elliott Williams and Yusuke Suzuki.

Internal SDKs for iOS 26 and macOS 26 enabled `-no_inits` by default for
targets that are included in the dyld shared cache, which broke the
Fuzzilli build.  This is fixed by adding linker flags to ignore static
initializers and to reassure the linker that the built product won't be
included in a dyld shared cache.

* Configurations/Sanitizers.xcconfig:
(WK_FUZZILLI_OTHER_LDFLAGS_YES):
- Add `-no_warn_inits` and `-not_for_dyld_shared_cache` to fix linking
  of JavaScriptCore.framework with Fuzzilli enabled.

Canonical link: https://commits.webkit.org/298304@main
